pkg/provider/aws/resources/rds.go

```diff
@@ -0,0 +1,327 @@
+		role.ManagedPolicies = append(role.ManagedPolicies, core.IaCValue{Resource: secretPolicy, Property: core.ARN_IAC_VALUE})
+		dag.AddDependency(secretPolicy, secret)
+
+		proxy = NewRdsProxy(orm, cfg.AppName, securityGroups, subnets, role, secret)
+		dag.AddDependency(proxy, secret)
+		proxyTargetGroup := NewRdsProxyTargetGroup(orm, cfg.AppName, instance, proxy)
+		dag.AddDependenciesReflect(secretVersion)
+		dag.AddDependenciesReflect(proxyTargetGroup)
+		dag.AddDependenciesReflect(proxy)
+		dag.AddDependenciesReflect(role)
+		dag.AddDependenciesReflect(secretPolicy)
+	}
+	dag.AddDependenciesReflect(instance)
+	dag.AddDependenciesReflect(subnetGroup)
+	return instance, proxy, nil
+}
+
+func (i *RdsInstance) GetConnectionPolicy(dag *core.ResourceGraph) *IamPolicy {
+	var pol *IamPolicy
+	upstreamDeps := dag.GetUpstreamDependencies(i)
+	for _, dep := range upstreamDeps {
+		if res, ok := dep.Source.(*IamPolicy); ok {
+			if len(res.Policy.Statement[0].Action) == 1 && res.Policy.Statement[0].Action[0] == "rds-db:connect" {
+				return res
+			}
+		}
+	}
+	return pol
+}
+
+// generateUsername generates a random username for the rds instance.
+//
+// The first letter of an RDS username must be a letter
+func generateUsername() string {
+	rand.Seed(time.Now().UnixNano())
+	initialChars := []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+		"abcdefghijklmnopqrstuvwxyz")
+	chars := []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+		"abcdefghijklmnopqrstuvwxyz" +
+		"0123456789")
+	length := 12
+	var b strings.Builder
+	b.WriteRune(chars[rand.Intn(len(initialChars))])
+	for i := 0; i < length-1; i++ {
+		b.WriteRune(chars[rand.Intn(len(chars))])
+	}
+	return b.String()
+}
+
+// generatePassword generates a random password for the rds instance.
+func generatePassword() string {
+	rand.Seed(time.Now().UnixNano())
+	chars := []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+		"abcdefghijklmnopqrstuvwxyz" +
+		"0123456789")
+	length := 16
+	var b strings.Builder
+	for i := 0; i < length; i++ {
+		b.WriteRune(chars[rand.Intn(len(chars))])
+	}
+	return b.String()
+}
+
+func NewRdsInstance(orm *core.Orm, appName string, subnetGroup *RdsSubnetGroup, securityGroups []*SecurityGroup) *RdsInstance {
+	return &RdsInstance{
+		Name:                             rdsInstanceSanitizer.Apply(fmt.Sprintf("%s-%s", appName, orm.ID)),
+		ConstructsRef:                    []core.AnnotationKey{orm.Provenance()},
+		SubnetGroup:                      subnetGroup,
+		SecurityGroups:                   securityGroups,
+		IamDatabaseAuthenticationEnabled: true,
+		DatabaseName:                     orm.ID,
+		Username:                         generateUsername(),
+		Password:                         generatePassword(),
+		Engine:                           "postgres",
+		EngineVersion:                    "13.7",
+		InstanceClass:                    "db.t4g.micro",
+		SkipFinalSnapshot:                true,
+		AllocatedStorage:                 20,
```

Comment:
Some of these feel like they should be configurable (e.g. `InstanceClass`), and pretty all of them (from `DatabaseName` down) feel like they belong in the IaC, not provider.

I know that means the IaC would need to handle `RdsInstance` specially, but I think that makes sense to do.

It may be useful to create a method on the graph, `replaceItem(V old, V new)` which replaces the vertex while keeping all the edges (having them instead point to the new item). So the iac plugin would take an `RdsInstance`, transform it into a `RenderableRdsInstance`, and replace that in the graph.

Not an absolute blocker, but if we don't do it now then I'd like at least a ticket to remind us to revisit it later, time permitting.

Reply:
yeah they should be, we just dont expose them yet. And why do you think they belong in the IaC? If they go in the klotho configuration then it would be handled by the provider and thats our layer of configuratbility.

We will need the replacement method at some point, i agree, but im not sure that i understand the reasoning for rds to follow a different flow than everything else

Reply:
eh, yeah maybe it's not. idk, these just feel lower level. But maybe I just need to adjust my mental model a bit.

Reply:
yeah i mean these could all affect things like cost, throughput, storage, etc which is information we will want to have within the provider to optimize later. Anything within the IaC is also static so if we want any configuratibility we cannot put these in the IaC. I realize they are static now but the intention is that they wont be forever

Reply:
Ah, that all makes sense — good call.